### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by Wind010**
- Organize common, client, and server specific code. 
- Add client and server Dockerfiles.  
- Add docker-compose.yaml.  
- Update documentation.

### CODE REVIEW

**Comment by DotNetRussell** [comment]
This looks awesome! I'll pull it down today and verify it today

Thank you

**Comment by DotNetRussell** [comment]
Did you run into this when you were trying to build the server image? 

Looks like it can't see the common directory 
image

**Comment by Wind010** [comment]
The docker build command has to be run from root of repository.  Let me
double check the documentation.

**Comment by Wind010** [comment]
> Did you run into this when you were trying to build the server image?
> 
> Looks like it can't see the common directory image

### Docker

All commands should be run from the root of the repository.
Build the client image:
```sh
docker build -f ./client/Dockerfile --build-arg ENCRYPTION_KEY=your_key --build-arg HOST=your_server_ip --build-arg PORT=5680 -t ensemble_client:latest .
```

Build the server image.  It's expected that you have created a `config.json` file similar to the Sample-Config.json
for your configuration in the `server` directory.  The server image build expects this:
```sh
docker build -f ./server/Dockerfile -t ensemble_server:latest .
```

image

Docker will only have access to subdirectories where it is run from so the `Dockerfile` has to be explicit path if it's in one.

File: server/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+COPY /server/.config.json .
+
+RUN pip3 install ${TRUSTED_PYPI_HOSTS} -r requirements.txt
+
+ADD ./* .
+
+RUN chmod +x ensemble_director.py
+
+RUN pip --no-cache-dir install -r requirements.txt
+CMD ["python3", "ensemble_director.py"]
```
Comment (DotNetRussell) [inline server/Dockerfile:52]: Can't seem to get this to work with the config file. Can you share what run command you're using? 

Seems like we need something like you have in the client dockerfile to pass the config file path in

Comment (Wind010) [reply server/Dockerfile:52]: It should just be this command run from the root of the repository. 
```sh
docker build -f ./client/Dockerfile --build-arg ENCRYPTION_KEY=your_key --build-arg HOST=your_server_ip --build-arg PORT=5680 -t ensemble_client:latest .
```

I can share the encryption key I used in a DM.

**Pull request merged**